src/worker/swirl/runner.rs

```diff
@@ -26,155 +23,153 @@ fn runnable<J: BackgroundJob>(ctx: &J::Context, payload: serde_json::Value) -> a
-        self.thread_pool.join();
-        let panic_count = self.thread_pool.panic_count();
-        if panic_count == 0 {
-            Ok(())
-        } else {
-            Err(anyhow!("{panic_count} threads panicked"))
-        }
+pub struct RunningRunner {
+    handles: Vec<JoinHandle<()>>,
+}
+
+impl RunningRunner {
+    /// Wait for all background workers to shut down.
+    pub fn wait_for_shutdown(self) {
+        self.handles.into_iter().for_each(|handle| {
+            if let Err(error) = handle.join() {
+                let error = try_to_extract_panic_info(&error);
+                warn!(%error, "Background worker thread panicked");
+            }
+        })
     }
 }
 
 struct Worker<Context> {
     connection_pool: DieselPool,
     context: Context,
     job_registry: JobRegistry<Context>,
-    sender: SyncSender<Event>,
+    shutdown_when_queue_empty: bool,
+    poll_interval: Duration,
 }
 
 impl<Context: Clone + Send + 'static> Worker<Context> {
-    fn run_next_job(&self) {
-        let conn = &mut *match self.connection_pool.get() {
-            Ok(conn) => conn,
-            Err(e) => {
-                // TODO: Review error handling and possibly drop all usage of `let _ =` in this file
-                let _ = self.sender.send(Event::FailedToAcquireConnection(e));
-                return;
-            }
-        };
-
-        let job_run_result = conn.transaction::<_, diesel::result::Error, _>(|conn| {
-            debug!("Looking for next background worker job…");
-            let job = match storage::find_next_unlocked_job(conn).optional() {
-                Ok(Some(j)) => {
-                    let _ = self.sender.send(Event::Working);
-                    j
+    /// Run background jobs forever, or until the queue is empty if `shutdown_when_queue_empty` is set.
+    pub fn run(&self) {
+        loop {
+            match self.run_next_job() {
+                Ok(Some(_)) => {}
+                Ok(None) if self.shutdown_when_queue_empty => {
+                    debug!("No pending background worker jobs found. Shutting down the worker…");
+                    break;
                 }
                 Ok(None) => {
-                    let _ = self.sender.send(Event::NoJobAvailable);
-                    return Ok(());
+                    debug!(
+                        "No pending background worker jobs found. Polling again in {:?}…",
+                        self.poll_interval
+                    );
+                    std::thread::sleep(self.poll_interval);
                 }
-                Err(e) => {
-                    let _ = self.sender.send(Event::ErrorLoadingJob(e));
-                    return Err(RollbackTransaction);
+                Err(error) => {
+                    error!(%error, "Failed to run job");
+                    std::thread::sleep(self.poll_interval);
                 }
+            }
+        }
+    }
+
+    /// Run the next job in the queue, if there is one.
```

Comment:
Let's document the return values here, since I spent more than one second thinking about it when reviewing the `match` above:

```suggestion
    /// Run the next job in the queue, if there is one.
    ///
    /// Returns:
    /// - `Ok(Some(job_id))` if a job was run successfully
    /// - `Ok(None)` if no jobs were waiting
    /// - `Err(...)` if there was an error retrieving the job, or the job returned an error
```